### PR TITLE
Patient menu fixes

### PR DESCRIPTION
**Pull request by matrix-amiel**
Hi.

In the our clinics we change the standard patient menu and remove part of the links, I found that in the new style of menu the ID's of links is numeric in ascending order and it is hardcode so there is problem with the mark as active of the current page.
 (For example - I remove History tab so issues is number 2 but the id is still 'nav-list3' therefore the  'Documenst' tab marks as active)
I changed it to the Id from the json and now it is dynamic.   

In additional I added option to add class to link (in the json), we need it because we need change the style of the links (we have extra modules and no space in the one line with current style)   

And more translation fix..

Thanks

Amiel (Matrix)

### CODE REVIEW

File: library/menu/src/PatientMenuRole.php [hunk 1]
```diff
@@ -181,14 +181,16 @@ public function displayHorizNavBarMenu()
                 // flatten to only show children items
                 foreach ($value->children as $children_key => $children_value) {
                     $link = ($children_value->pid != "true") ? $children_value->url : $children_value->url . attr($pid);
-                    $list = '<li class="oe-bold-black" id="nav-list'. $li_id.'">';
+                    $class = isset($children_value->class) ? $children_value->class : '';
+                    $list = '<li class="oe-bold-black ' . $class . '" id="' . $children_value->menu_id . '">';
```
Comment (bradymiller) [inline library/menu/src/PatientMenuRole.php]: should place attr() around both the above variables

File: library/menu/src/PatientMenuRole.php [hunk 2]
```diff
@@ -181,14 +181,16 @@ public function displayHorizNavBarMenu()
...
                     $list .= '<a href="' . $link . '" onclick="' . $children_value->on_click .'"> ' . text($children_value->label) . ' </a>';
                     $list .= '</li>';
                 }
                 echo $list. "\r\n";
             } else {
                 $link = ($value->pid != "true") ? $value->url : $value->url . attr($pid);
-                $list = '<li class="oe-bold-black"id="nav-list'. $li_id.'">';
+                $class = isset($value->class) ? $value->class : '';
+                $list = '<li class="oe-bold-black ' . $class . '" id="' . $value->menu_id . '">';
```
Comment (bradymiller) [inline library/menu/src/PatientMenuRole.php]: should place attr() around both the above variables

File: library/options.inc.php [hunk 3]
```diff
@@ -2036,7 +2036,7 @@ function generate_display_field($frow, $currvalue)
         }
         // If match is not found in main and backup lists, return the key with exclamation mark
         if ($s == '') {
-            $s = nl2br(htmlspecialchars($currvalue, ENT_NOQUOTES)).
+            $s = nl2br(xlt($currvalue)).
```
Comment (bradymiller) [inline library/options.inc.php]: Isn't this only showing if there is an error? (ie. the entry is missing)

Comment (matrix-amiel) [reply library/options.inc.php]: Right. It also prints the old value in the case the row was removed, but it is not translated..

Comment (matrix-amiel) [reply library/options.inc.php]: image

image

File: library/menu/src/PatientMenuRole.php (hunk 1 above)
Comment (matrix-amiel) [reply library/menu/src/PatientMenuRole.php]: Done

File: library/menu/src/PatientMenuRole.php (hunk 2 above)
Comment (matrix-amiel) [reply library/menu/src/PatientMenuRole.php]: Done

File: library/menu/src/PatientMenuRole.php [hunk 4]
```diff
@@ -181,15 +181,17 @@ public function displayHorizNavBarMenu()
                 // flatten to only show children items
                 foreach ($value->children as $children_key => $children_value) {
                     $link = ($children_value->pid != "true") ? $children_value->url : $children_value->url . attr($pid);
-                    $list = '<li class="oe-bold-black" id="nav-list'. $li_id.'">';
-                    $list .= '<a href="' . $link . '" onclick="' . $children_value->on_click .'"> ' . text($children_value->label) . ' </a>';
+                    $class = isset($children_value->class) ? $children_value->class : '';
+                    $list = '<li class="oe-bold-black ' . attr($class) . '" id="' . attr($children_value->menu_id) . '">';
+                    $list .= '<a href="' . attr($link) . '" onclick="' . attr($children_value->on_click) .'"> ' . text($children_value->label) . ' </a>';
```
Comment (bradymiller) [inline library/menu/src/PatientMenuRole.php]: oops, lets remove the attr from the $children_value->on_click entry (that can break things)

File: library/menu/src/PatientMenuRole.php [hunk 5]
```diff
@@ -181,15 +181,17 @@ public function displayHorizNavBarMenu()
...
                     $list .= '</li>';
                 }
                 echo $list. "\r\n";
             } else {
                 $link = ($value->pid != "true") ? $value->url : $value->url . attr($pid);
-                $list = '<li class="oe-bold-black"id="nav-list'. $li_id.'">';
-                $list .= '<a href="' . $link . '" onclick="' . $value->on_click .'"> ' . text($value->label) . ' </a>';
+                $class = isset($value->class) ? $value->class : '';
+                $list = '<li class="oe-bold-black ' . attr($class) . '" id="' . attr($value->menu_id) . '">';
+                $list .= '<a href="' . attr($link) . '" onclick="' . attr($value->on_click) .'"> ' . text($value->label) . ' </a>';
```
Comment (bradymiller) [inline library/menu/src/PatientMenuRole.php]: oops, lets remove the attr from the $children_value->on_click entry (that can break things)

File: library/options.inc.php (hunk 3 above)
Comment (bradymiller) [reply library/options.inc.php]: do the following instead:
```
xl_list_label
$s = nl2br(text(xl_list_label($currvalue))).
```

Comment (bradymiller) [reply library/options.inc.php]: To explain above xl_list_label relies on a option in globals to translate lists (it is on by default but gives folks option to turn off): https://www.open-emr.org/wiki/index.php/Administration_Globals#Translate_Lists

Comment (matrix-amiel) [reply library/options.inc.php]: fixed

File: library/options.inc.php [hunk 6]
```diff
@@ -2036,7 +2036,7 @@ function generate_display_field($frow, $currvalue)
         }
         // If match is not found in main and backup lists, return the key with exclamation mark
         if ($s == '') {
-            $s = nl2br(htmlspecialchars($currvalue, ENT_NOQUOTES)).
+            $s = nl2br(text(xl_list_label($currvalue), ENT_NOQUOTES)).
```
Comment (bradymiller) [inline library/options.inc.php]: so close :) (need to remove the `, ENT_NOQUOTES`)
```
$s = nl2br(text(xl_list_label($currvalue))).
```

Comment (matrix-amiel) [reply library/options.inc.php]: oops..
Done

**Comment by bradymiller** [comment]
hi @matrix-amiel , Code looks good and will bring in. Thanks for the fixes! -brady

**Pull request merged**